Electric Heating Coil Inlet and Outlet Air Conditions and Sensible Capacity Tabular Reporting Issue for Four Pipe FCU

Nigusse opened the pull request:
Pull request overview
---------------------
 - Fixes #9802 
 - Fixes incorrect electric heating coil entering air conditions and sensible heating capacity reported to "Coil Sizing Summary" tabular outputs. 

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

Nigusse committed:
Initial Fix of FanCoil Elec HCoil Sizing

Nigusse committed:
Merge branch 'develop' into Issue9802_4PipeFCU_SAT_Reporting

amirroth commented on src/EnergyPlus/ReportCoilSelection.cc:
```diff
@@ -1675,7 +1675,14 @@ void ReportCoilSelection::setCoilHeatingCapacity(
         } else if (state.dataSize->ZoneEqFanCoil) {
             if (c->coilDesEntTemp == -999.0) { // don't overwrite if already set directly by setCoilEntAirTemp
-                Real64 desOAFlowFrac = state.dataSize->FinalZoneSizing(curZoneEqNum).DesHeatOAFlowFrac;
+                Real64 desOAFlowFrac = 0.0;
```
Much opportunity to use `auto &` shortcuts in this module there is.

Nigusse replied:
I agree. I will get back to it when the functional changes are complete.

Nigusse committed:
Added Unit Test

Nigusse committed:
Applied ShortCuts and Cleanup

Nigusse commented:
The issues was caused by `0.0` outdoor air fraction for sensible heating capacity calculation in the `ReportCoilSelection.cc` module, when there is outdoor air specified. Modified the code to use the correct outdoor fraction, which is the same as  the value used for sizing the parent object heating capacity, resolved the problem. Now the total and sensible heating capacity do match. See below a snap shot of the tabular output of the  `Coil Sizing Summary` table. Note that the coil entering and leaving air conditions have also changed.

![TabularOutput-2](https://user-images.githubusercontent.com/5805126/216129491-382e3d81-078e-4dbd-bf78-6098909bcfcb.png)

Nigusse commented on src/EnergyPlus/ReportCoilSelection.cc:
```diff
@@ -1607,102 +1609,105 @@ void ReportCoilSelection::setCoilHeatingCapacity(
+                    desOAFlowFrac = std::min(state.dataEnvrn->StdRhoAir * zoneEqSizing.OAVolFlow / finalZoneSizing.DesHeatMassFlow, 1.0);
+                } else {
+                    desOAFlowFrac = finalZoneSizing.DesHeatOAFlowFrac;
+                }
```
The above code is the fix for this issue. The code `zoneEqSizing.OAVolFlow > 0.0 && finalZoneSizing.DesHeatMassFlow > 0.0` condition is the new code added to calculate the `OA` fraction if the `OAVolFlow > 0.0` else it defaults to the existing code.  The other changes else where are cleanup only.

Nigusse committed:
Replaced Bad ShortCuts

Nigusse committed:
Merge branch 'develop' into Issue9802_4PipeFCU_SAT_Reporting

Nigusse commented on tst/EnergyPlus/unit/ReportCoilSelection.unit.cc:
```diff
@@ -802,3 +802,145 @@ TEST_F(EnergyPlusFixture, ReportCoilSelection_ZoneEqCoolingCoil)
+    EXPECT_NEAR(RatedCoilTotCap, c1->coilTotCapAtPeak, 0.1);
+    EXPECT_NEAR(RatedCoilTotCap, c1->coilSensCapAtPeak, 0.1);
+    EXPECT_NEAR(result_sensCapacity, c1->coilSensCapAtPeak, 0.1);
+}
```
This unit tests demonstrates the correct coil leaving and design air conditions. Also the coil heating total and sensible capacity are calculated correctly.

Nigusse commented on src/EnergyPlus/ReportCoilSelection.cc:
```diff
@@ -1732,27 +1730,33 @@ void ReportCoilSelection::setCoilHeatingCapacity(
+                        c->coilDesVolFlow = unitarySysEqSizing.AirVolFlow;
+                    } else if (unitarySysEqSizing.HeatingAirFlow) {
+                        c->coilDesVolFlow = unitarySysEqSizing.HeatingAirVolFlow;
+                    }
```
Most of the above code changes are `ShortCuts` applied for readability.

Nigusse reviewed (commented):
Code change Walk-through update. This PR is ready for review.

Nigusse committed:
Merge branch 'develop' into Issue9802_4PipeFCU_SAT_Reporting

Myoldmopar commented:
Confirmed the defect file shows the electric heating coil with bad sensible/total capacity pair, and with this branch they match, as one would expect.  Everything builds fine with latest develop pulled in and tests pass locally.  CI shows lots of files with coil sizing diffs as the only change in tabular output, no math diffs.  Merging this, thanks @Nigusse.
